### PR TITLE
4891 bulk export do not recurse unasked for resources

### LINKED ISSUE #4891: Batch2 Export recurses on resources that are not requested
Submitting a batch2 job with a _type filter that doesn't include (for example) Practitioners, but does include Encounters, will return practitioners that are linked to any found Encounters.

These should not be returned

### PR DESCRIPTION
**Pull request by TipzCM**
closes #4891

### CODE REVIEW

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/bulk/export/svc/JpaBulkExportProcessor.java [hunk 1]
```diff
@@ -443,14 +464,16 @@ private void extract(List<MdmPidTuple<JpaPid>> theGoldenPidTargetPidTuples, Map<
...
+																			 List<JpaPid> thePatientPids,
 																			 ExportPIDIteratorParameters theParams,
-																			 RuntimeResourceDefinition theDef) {
+																			 RuntimeResourceDefinition theDef) throws IOException {
 
 		//Convert Resource Persistent IDs to actual client IDs.
-		Set<JpaPid> pidSet = new HashSet<>(JpaPidChunk);
-		Set<String> resourceIds = myIdHelperService.translatePidsToFhirResourceIds(pidSet);
+		Set<JpaPid> pidSet = new HashSet<>(thePatientPids);
+		Set<String> patientId = myIdHelperService.translatePidsToFhirResourceIds(pidSet);
```
Comment (lukedegruchy) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/bulk/export/svc/JpaBulkExportProcessor.java:476]: Nitpick:  patientIds (plural)

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/bulk/export/svc/JpaBulkExportProcessor.java [hunk 2]
```diff
@@ -461,29 +484,49 @@ private void queryResourceTypeWithReferencesToPatients(Set<JpaPid> theReadPids,
...
+				partitionId)) {
+				while (resultIterator.hasNext()) {
+					theReadPids.add(resultIterator.next());
+				}
+			}
+
+			// Construct our Includes filter
+			// We use this to recursively fetch resources of interest
+			// (but should only request those the user has requested/can see)
+			Set<Include> includes = new HashSet<>();
```
Comment (lukedegruchy) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/bulk/export/svc/JpaBulkExportProcessor.java:512]: Nitpick:  Why not a Stream map()?

**Review by lukedegruchy — APPROVED** [review]

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/bulk/export/svc/JpaBulkExportProcessor.java [hunk 3]
```diff
@@ -219,23 +224,24 @@ private LinkedHashSet<JpaPid> getPidsForSystemStyleExport(ExportPIDIteratorParam
 			Logs.getBatchTroubleshootingLog().debug("Executing query for bulk export job[{}] chunk[{}]: {}", theJobId, theChunkId, map.toNormalizedQueryString(myContext));
 
 			// requires a transaction
-			IResultIterator<JpaPid> resultIterator = searchBuilder.createQuery(map,
+			try (IResultIterator<JpaPid> resultIterator = searchBuilder.createQuery(map,
```
[the remark this answers is not in the document]
Comment (TipzCM) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/bulk/export/svc/JpaBulkExportProcessor.java:227]: Just fixing cause i was there.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/bulk/export/svc/JpaBulkExportProcessor.java (hunk 1 above)
Comment (TipzCM) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/bulk/export/svc/JpaBulkExportProcessor.java:476]: changed

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/bulk/export/svc/JpaBulkExportProcessor.java (hunk 2 above)
Comment (TipzCM) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/bulk/export/svc/JpaBulkExportProcessor.java:512]: I prefer for loops.

Unless there's a really good reason not to use a for loop, i always use them.

**Pull request merged**